Custom authentication support (#6162)
strimzi/strimzi-kafka-operator: documentation/api/io.strimzi.api.kafka.model.authentication.KafkaClientAuthenticationCustom.adoc

> To configure custom authentication, set the `type` property to `custom`.

Reviewer: Not sure we need this file. If I understand correctly, this is custom Kafka listener/authentication option. In which case, this be added under the `authentication` heading in _documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc_

Author: Woops, that was an accident on my behalf! I renamed + linked from `documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc`

> Custom authentication allows for any type of kafka supported authentication to be used.

Reviewer: ```suggestion
Custom authentication allows for any type of kafka-supported authentication to be used.
```

Reviewer 2: `any type` sounds a bit weird. Should it be more about `custom type`?

            oauthbearer.sasl.server.callback.handler.class: server.class
            oauthbearer.sasl.login.callback.handler.class: login.class
            oauthbearer.connections.max.reauth.ms: 999999999
            sasl.enabled.mechanisms: oauthbearer
            oauthbearer.sasl.jaas.config: |
              org.apache.kafka.common.security.oauthbearer.OAuthBearerLoginModule required ;
          secrets:
            - name: example
----

.Protocol Map
The protocol map that is generated for Strimzi uses the above `sasl` and `tls` fields to determine which protocol to map to the listener.

* SASL = True, TLS = True -> SASL_SSL
* SASL = False, TLS = True -> SSL
* SASL = True, TLS = False -> SASL_PLAINTEXT
* SASL = False, TLS = False -> PLAINTEXT

.Listener Config
All configuration listed under `listener-config` will be prefixed with `listener.name.<listener-name>-<port>`.
Therefore, the above `sasl.enabled.mechanisms` becomes `listener.name.<listener-name>-<port>.sasl.enabled.mechanisms`.

.Secrets
`Secrets` will be mounted to `/opt/palantir/custom-auth-secrets/custom-<listener-name>-<port>/<secret-name>` in the Kafka broker nodes' containers.

> For example, the mounted secret `example` in the above setup would be located at `/opt/palantir/custom-auth-secrets/custom-oauth-bespoke-9093/example`.

Reviewer 2: I don't think the paths correspond to the code (+ even if I read the code wrong, should be `/opt/kafka` ;-))

Author: woops! my apologies, silly mistake on my behalf :)

.Principal
A custom principal builder can be set in the general Kafka cluster config.

However, please note while it is possible to set your own principal builder, there are some requirements:

. The specified principal builder class exists on the image.
If you're using the open source Strimzi images, you'll need to rebuild your own with your desired classes.
*Note --* It's worth opening a Github issue on this particular problem _before_ building your own, as one may already exist.
. No other listener is using `oauth` type authentication.
This is because the Oauth listener forcefully appends its own principlar builder to the kafka config.
. The specified principal builder is compatible with Strimzi.

Custom principal builders need to support peer certificates for authentication, as Strimzi uses this to manage the Kafka cluster.

> This is expected to be identical, or very similar to https://github.com/strimzi/strimzi-kafka-oauth/blob/main/oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipalBuilder.java[Strimzi's OAuth library].

Reviewer: ```suggestion
ifdef::Downloading[]
A custom OAuth principal builder might be identical or very similar to the Strimzi https://github.com/strimzi/strimzi-kafka-oauth/blob/main/oauth-server/src/main/java/io/strimzi/kafka/oauth/server/OAuthKafkaPrincipalBuilder.java[OAuth principal builder].
endif::Downloading[]
```

Reviewer: I've added a condition here, which we normally do with links to github repo

In particular, it leverages https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java#L73-L79[Kafka's default principal builder class], which allows the building of a principal based upon the name of the of peer certificate.

> Thus, it is expected that the custom principal builder provides a principal of type "user" with the name being that of the ssl peer certificate.

Reviewer: ```suggestion
The custom principal builder should provide a principal of type "user" using the name of the SSL peer certificate.
```

> An example of such a principal builder which satisfies Strimzi’s auth requirements:

Reviewer: ```suggestion
The following example shows a custom principal builder that satisfies the OAuth requirements of Strimzi.
```

> [source,java,subs="attributes+"]

Reviewer: ```suggestion
.Example principal builder for custom OAuth configuration
[source,java,subs="attributes+"]
```
